File: back/DevC/src/main/java/com/back/devc/domain/post/post/dto/PostDetailResponse.java
package com.back.devc.domain.post.post.dto;

import com.back.devc.domain.member.member.util.MemberDisplayUtil;
import com.back.devc.domain.post.post.entity.Post;

import java.time.LocalDateTime;

public record PostDetailResponse(
        Long postId,
        String title,
        String content,
        Long userId,
        String writerName,
        Long categoryId,
        int viewCount,
        int likeCount,
        int commentCount,
        // 게시글 상세 화면에서 북마크 수를 바로 표시하기 위해 함께 내려주는 값
        int bookmarkCount,
        // 현재 로그인 사용자가 이 게시글을 북마크했는지 여부
        boolean bookmarked,
        boolean liked,
        LocalDateTime createdAt,
        LocalDateTime updatedAt
) {

    public static PostDetailResponse from(Post post) {
        return from(post, false, false, 0);
    }

    public static PostDetailResponse from(Post post, boolean liked) {
        return from(post, liked, false, 0);
    }

    /**
     * 게시글 상세조회 응답 생성
     */
    public static PostDetailResponse from(Post post, boolean liked, boolean bookmarked, int bookmarkCount) {
        return new PostDetailResponse(
                post.getPostId(),
                post.getTitle(),
                post.getContent(),
                post.getMember() != null ? post.getMember().getUserId() : null,
                MemberDisplayUtil.getDisplayName(post.getMember()),
                post.getCategory().getCategoryId(),
                post.getViewCount(),
                post.getLikeCount(),
                post.getCommentCount(),
                bookmarkCount,
                bookmarked,
                liked,
                post.getCreatedAt(),
                post.getUpdatedAt()
        );
    }
}

File: back/DevC/src/main/kotlin/com/back/devc/domain/post/post/dto/PostDetailResponse.kt
package com.back.devc.domain.post.post.dto

import com.back.devc.domain.member.member.util.MemberDisplayUtil
import com.back.devc.domain.post.post.entity.Post
import java.time.LocalDateTime

data class PostDetailResponse(
    val postId: Long,
    val title: String,
    val content: String,
    val userId: Long?,
    val writerName: String,
    val categoryId: Long,
    val viewCount: Int,
    val likeCount: Int,
    val commentCount: Int,
    val bookmarkCount: Int,
    val bookmarked: Boolean,
    val liked: Boolean,
    val createdAt: LocalDateTime,
    val updatedAt: LocalDateTime
) {
    companion object {
        @JvmStatic
        fun from(
            post: Post,
            liked: Boolean = false,
            bookmarked: Boolean = false,
            bookmarkCount: Int = 0
        ): PostDetailResponse {
            return PostDetailResponse(
                postId = post.postId
                    ?: throw IllegalStateException("Post ID cannot be null"),
                title = post.title,
                content = post.content,
                userId = post.member?.userId,
                writerName = MemberDisplayUtil.getDisplayName(post.member),
                categoryId = post.category.categoryId,
                viewCount = post.viewCount,
                likeCount = post.likeCount,
                commentCount = post.commentCount,
                bookmarkCount = bookmarkCount,
                bookmarked = bookmarked,
                liked = liked,
                createdAt = post.createdAt,
                updatedAt = post.updatedAt
            )
        }
    }
}